# nf-transFlow

A Nextflow (DSL2) pipeline designed for comprehensive transcriptome analysis including quality assessment, pre-processing, _de novo_ assembly, and functional annotation. This pipeline is specifically optimized for deployment on high-performance computing (HPC) clusters and executed _via_ Slurm Workload Manager.

## Key Features

- **Transcriptome Quality Assessment**: Ensures high-quality data processing with integrated quality checks.
- **Pre-processing**: Streamlines the handling of RNA-seq data to prepare for assembly.
- **De Novo Assembly**: Assembles transcripts from RNA-seq data without the need for a reference genome.
- **Functional Annotation**: Annotates assembled transcripts to provide insights into gene function and expression.
- **HPC Optimization**: Designed to efficiently utilize HPC resources for scalable and fast transcriptome analysis.

## Dependecies 

To use this pipeline you will need:
- Nextflow >= 23.04.1
- Singularity >= 3.8.7

## Installation and deployment

1. Install the nf-transFlow pipeline resources into your HPC cluster directory where you will be performing the analysis:  

```
git clone https://github.com/krishnan-Rama/transFlow.git
```

2. Add the RNA-seq fastq reads ({_1,_2,_R1,_R2}.fastq.gz format) in the `raw_data` directory.  

3. Change module versions (if applicable) for singularity and nextflow (prerequisites) in `run_nextflow.sh` script accordingly as available on your HPC cluster, if necessary.

4.  Run the pipeline by executing the command `./deploy.sh`  

6. The prompt will ask you to enter your preferred HPC partition name to submit the job and the species identifier (e.g. Hsap for _Homo sapiens_), simply type the name and return.

**Results location:** The results for each step in the pipeleine will be in the **`./output`** directory.   

### Workflow diagram with pipeline characteristics
![workflow](https://github.com/krishnan-Rama/transpipeline_containerised/assets/104147619/892ae381-69b3-45e8-a485-ccd50cf1794a)

### Auto-generated transFlow flowchart
![workflow](https://github.com/krishnan-Rama/transpipeline_containerised/blob/302bdd18135b8ac3de86a915e91df5e724d28d13/autogenerated_workflow.png)

### Concluding Note

To successfully run this pipeline, ensure the fastq reads are in correct format and HPC partition name is correct. For more detailed information or to modify specific pipeline behaviours, refer to individual nextflow scripts in the `modules` directory and the configuration file.

### Author Information

Rama Krishnan - krishnanr1@cardiff.ac.uk
